Introduce grafana callback plugin.

rrey opened the pull request:
##### SUMMARY

The grafana plugin plublishes annotations to the HTTP API available in
Grafana 4.6+

The callback publishes:
* An annotation on playbook start, tagged with "ansible",
 "ansible_event_start" and playbook name, example:

```
{
    "text": "Started playbook test.yml\n\nFrom 'pc45.home'\nBy user 'remirey'\n",
    "tags": ["ansible", "ansible_event_start", "test.yml"],
    "time": 1514291163000
}
```

* An annotation on error containing the host and task who failed and
  tagged with "ansible", "ansible_event_failure" and playbook name, example:

```
{
    "text": "Playbook test.yml Failure !\n\nFrom 'pc45.home'\nBy user 'remirey'\n\n'TASK: simulate failure' failed on localhost\n\ndebug: {\"changed\": false, \"msg\": \"Some random failure\"}\n",
    "tags": ["ansible", "ansible_event_failure", "test.yml"],
    "time": 1514291165000
}
```

* A region annotation emitted on playbook stats, tagged with "ansible",
  "ansible_report" and playbook name, example:

```
{
    "text": "Playbook test.yml\nDuration: 1.641703\nStatus: FAILED\n\nFrom 'pc45.home'\nBy user 'remirey'\n\nResult:\n{\"localhost\": {\"unreachable\": 0, \"skipped\": 0, \"ok\": 2, \"changed\": 1, \"failures\": 1}}\n",
    "tags": ["ansible", "ansible_report", "test.yml"],
    "isRegion": true,
    "timeEnd": 1514291165000,
    "time": 1514291163000
}
```

##### ISSUE TYPE

 - Feature Pull Request

##### COMPONENT NAME

plugins/callback/grafana.py

##### ANSIBLE VERSION

```
devel
```

##### ADDITIONAL INFORMATION

Fixes #34225

rrey committed:
Introduce grafana callback plugin.

The grafana plugin plublishes annotations to the HTTP API available in
Grafana 4.6+

The callback publishes:
* An annotation on playbook start, tagged with "ansible",
 "ansible_event_start" and playbook name, example:

```
{
    "text": "Started playbook test.yml\n\nFrom 'pc45.home'\nBy user 'remirey'\n",
    "tags": ["ansible", "ansible_event_start", "test.yml"],
    "time": 1514291163000
}
```

* An annotation on error containing the host and task who failed and
  tagged with "ansible", "ansible_event_failure" and playbook name, example:

```
{
    "text": "Playbook test.yml Failure !\n\nFrom 'pc45.home'\nBy user 'remirey'\n\n'TASK: simulate failure' failed on localhost\n\ndebug: {\"changed\": false, \"msg\": \"Some random failure\"}\n",
    "tags": ["ansible", "ansible_event_failure", "test.yml"],
    "time": 1514291165000
}
```

* A region annotation emitted on playbook stats, tagged with "ansible",
  "ansible_report" and playbook name, example:

```
{
    "text": "Playbook test.yml\nDuration: 1.641703\nStatus: FAILED\n\nFrom 'pc45.home'\nBy user 'remirey'\n\nResult:\n{\"localhost\": {\"unreachable\": 0, \"skipped\": 0, \"ok\": 2, \"changed\": 1, \"failures\": 1}}\n",
    "tags": ["ansible", "ansible_report", "test.yml"],
    "isRegion": true,
    "timeEnd": 1514291165000,
    "time": 1514291163000
}
```

Fixes #34225

rrey commented:
+label new_plugin

rrey commented:
Here is a screenshot of the result with the region annotation hovered:
![capture](https://user-images.githubusercontent.com/2752379/34462225-56b254d6-ee3f-11e7-9c74-67a3ec08b52f.png)

rrey committed:
Add support of HTTPS on grafana server side.

rrey committed:
Add a warning when grafana server does not answer with HTTP 200.

To avoid silent callback failures, adding a warning message informing of the
issue.

rrey commented:
Hi guys, 

(Happy New Year !)
bot_broken with following error:
```
2018-01-01 13:15:06 ERROR: Command "ssh -tt -q -i test/cache/id_rsa -o BatchMode=yes -o ServerAliveCountMax=4 -o ServerAliveInterval=15 -o StrictHostKeyChecking=no -o UserKnownHostsFile=/dev/null -p 22 ec2-user@ec2-35-153-143-31.compute-1.amazonaws.com su -l root -c ''"'"'chmod +x /tmp/remote.sh && /tmp/remote.sh freebsd'"'"''" returned exit status 255.
###
```
Unless I'm missing something, it is not related to my patch, I relaunched the tests several times but I keep on having this error.

mattclay commented:
@rrey Try rebasing your branch to see if that resolves the CI failure.

rrey commented:
@mattclay, thanks for the feedback. I did the rebase and re-pushed. 
Bot failed again on a failed ssh connection attempt on Azure:

```
2018-01-08 09:02:52 ERROR: Command "ssh -tt -q -i test/cache/id_rsa -o BatchMode=yes -o ServerAliveCountMax=4 -o ServerAliveInterval=15 -o StrictHostKeyChecking=no -o UserKnownHostsFile=/dev/null -p 22 azure@52.151.26.140 sudo -in bash -c ''"'"'ansible/test/runner/test.py integration -v --retry-on-error posix/ci/group1/ --exclude posix/ci/cloud/ --remote-terminate always --remote-stage prod --remote-provider default --exclude posix/ci/cloud/ --exclude skip/rhel/ --metadata metadata-b97YBr.json --color yes --requirements --allow-destructive'"'"''" returned exit status 1.
```

Regards

mattclay commented:
Tests are now passing.

rrey committed:
Rename grafana.py to grafana_annotations.py

mattclay commented:
CI timed out. I've restarted it.

daf committed:
Parse $GRAFANA_SECURE as an integer

rrey committed:
Add documentation

rrey commented:
@mattclay , is there anything wrong with the PR that would explain that I didn't have any review yet ?

rrey commented:
!new_plugin

seuf commented on lib/ansible/plugins/callback/grafana_annotations.py:
```diff
@@ -0,0 +1,203 @@
+from datetime import datetime
+
+try:
+    import httplib
```
you can use from `ansible.module_utils.urls import fetch_url` instead of import additional library

rrey replied:
The module_utils are for modules, here I'm submitting a callback.
The fetch_url function requires the module as first argument, I could try cheating by adding the params dict in my callback class, but the code also expects a fail_json method the is used to report the module failure.
That would be ugly imo.

seuf commented on lib/ansible/plugins/callback/grafana_annotations.py:
```diff
@@ -0,0 +1,203 @@
+            self.disabled = True
+            self._display.warning("GRAFANA_API_TOKEN not defined in the environment")
+        self.grafana_server = os.getenv('GRAFANA_SERVER', 'localhost')
+        self.grafana_port = int(os.getenv('GRAFANA_PORT', 3000))
```
why not use only one variable `grafana_url` like https://grafana-server:3000 instead of 3 distinct variable host / port and secure

rrey replied:
I'll repost with only 2, merging grafana_server and grafana_port. I'll keep the secure one mainly to avoid parsing the string looking for http or https.

seuf commented on lib/ansible/plugins/callback/grafana_annotations.py:
```diff
@@ -0,0 +1,203 @@
+    def __init__(self):
+        super(CallbackModule, self).__init__()
+
+        token = os.getenv('GRAFANA_API_TOKEN')
```
I suggest to rename this to `grafana_api_key` and also add `grafana_user` and `grafana_password` for users who don't want to create an api key for each organisation. (see  [grafana_datasource](https://github.com/ansible/ansible/blob/devel/lib/ansible/modules/monitoring/grafana_datasource.py) plugin for example)

rrey replied:
Ok.

seuf commented on lib/ansible/plugins/callback/grafana_annotations.py:
```diff
@@ -0,0 +1,203 @@
+DOCUMENTATION = '''
+    callback: grafana_annotations
+    type: notification
+    short_description: Sends events to Grafana HTTP API
```
send ansible events as annotations on charts to grafana over http api

rrey replied:
done

seuf commented on lib/ansible/plugins/callback/grafana_annotations.py:
```diff
@@ -0,0 +1,203 @@
+        text = PLAYBOOK_START_TXT.format(playbook=self.playbook,
+                                         hostname=self.hostname, username=self.username)
+        data = {
+            'time': to_millis(self.start_time),
```
`dashboardId` parameter is missing.
How do you specify on which dashboard annotations are created ?

rrey replied:
I did not put it on purpose because I felt that the annotations needs to be globals to allow one to display them on any dashboard.
But I can still add the feature allowing to provide a dashboard id.

rrey replied:
and I'll add panelId too.

seuf commented on lib/ansible/plugins/callback/grafana_annotations.py:
```diff
@@ -0,0 +1,203 @@
+                                         status=status, username=self.username,
+                                         summary=json.dumps(summarize_stat))
+
+        data = {
```
dashboardId parameter is missing.

rrey replied:
now done line 207

seuf commented on lib/ansible/plugins/callback/grafana_annotations.py:
```diff
@@ -0,0 +1,203 @@
+        text = PLAYBOOK_ERROR_TXT.format(playbook=self.playbook, hostname=self.hostname,
+                                         username=self.username, task=result._task,
+                                         host=result._host.name, result=self._dump_results(result._result))
+        data = {
```
dashboardId parameter is missing.

rrey replied:
now done line 207

seuf reviewed (changes requested):
Hi, I'm new contributor on ansible (with some grafana modules too). I've made some suggestions to your pull requests.

rrey commented:
@seuf, thanks for the review, the PR was quite silent :) I'll have a look at it this weekend.

rrey committed:
Add basic auth, dashboard id and panel id support.

Also get rid of GRAFANA_PORT to have it with server ip as a string and now
named GRAFANA_HOST.

rrey commented:
!ready_for_review

seuf commented on lib/ansible/plugins/callback/grafana_annotations.py:
```diff
@@ -0,0 +1,221 @@
+      - This callback will report start, failed and stats events to Grafana as annotations (https://grafana.com)
+    version_added: "2.5"
+    requirements:
+      - whitelisting in configuration
```
please add the httplib in requirements

rrey replied:
I think that the requirements section should only specify an external module requirement, not a standard one, right ?

bcoca replied:
correct, in any case i would advise using the open_url function that Ansible provides as it deals with many issues that httplib/urllib don't (versions, proxy, ssl, etc) and I would add a couple of options to allow users to deal with custom setups (validate_certs, etc).

seuf replied:
like @bcoca I suggest to use the Ansible provided functions instead of adding more dependencies..

rrey replied:
I will, Wagener I’ll Be back from my ski vacations ;)

seuf reviewed (changes requested):
Just a minor change in requirements, and it should be ok for me.

rrey committed:
Implement the use of open_url function in place of httplib

The use of open_url introduces new environment variables allowing to use
features from open_url:

 * GRAFANA_VALIDATE_CERT: will allow to enable/disable the certificate
   validation when using ssl.
 * HTTP_AGENT : will allow to change the 'User-agent' header value.

rrey commented:
Sorry for the wrong push ...
I did a rebase as my branch started to be old compared to the current devel.
I implemented the use of open_url but I did not implemented the client certificate usage.

It is a simple change to do but I'd like to have the CI testing it properly on the original project (https://github.com/rrey/ansible-callback-grafana-annotations) before pushing it to the ansible project.

rrey commented:
!bot_broken I think, the CI failure is not related to my patch.

mattclay commented:
@rrey I've restarted CI. In the future, if you see an unrelated CI failure, pushing an additional commit will trigger CI to run again. You can also request a contributor restart CI for you. Asking in #ansible-devel on IRC is often the quickest way to do that.

Flagging the PR as "bot broken" simply tells the bot to ignore the PR. The bot is only reporting CI failures -- it does't have anything to do with actually running the tests.

sivel commented on lib/ansible/plugins/callback/grafana_annotations.py:
```diff
@@ -0,0 +1,234 @@
+        self.grafana_user = os.getenv('GRAFANA_USER', None)
+        self.grafana_password = os.getenv('GRAFANA_PASSWORD', None)
+        self.dashboard_id = os.getenv('GRAFANA_DASHBOARD_ID', None)
+        self.panel_id = os.getenv('GRAFANA_PANEL_ID', None)
```
These `os.getenv` calls should likely be updated to use `self.get_option` and potentially `self.set_options`.  See the `slack` callback for examples.

sivel replied:
Without this change, the `DOCUMENTATION` isn't really being used, and is likely to cause confusion with users who assume they can set these like any other ansible config value.

rrey replied:
Hi @sivel, thanks for the review.
After a quick read of the code, I'm not sure to fully understand the get_option and DOCUMENTATION relation. 

From the slack example below:
```
    options:
      webhook_url:
        required: True
        description: Slack Webhook URL
        env:
          - name: SLACK_WEBHOOK_URL
        ini:
          - section: callback_slack
            key: webhook_url
```
Is the DOCUMENTATION string parsed to retrieve the env variable name that can contain the value and/or the ansible.cfg section and parameter key ?

Is there a clear documentation explaining this that I would have missed ?

sivel replied:
Yes, DOCUMENTATION is parsed to define configuration values for your plugin.  Any values supplied via ansible.cfg or env var will be made available to your plugin through the `set_options`/`get_option` API.

The slack callback makes use of both `set_options` and `get_option`

sivel commented on lib/ansible/plugins/callback/grafana_annotations.py:
```diff
@@ -0,0 +1,234 @@
+    return int(dt.strftime('%s')) * 1000
+
+
+def str2bool(string):
```
When you switch the callback to use the `get_option` API. This function can go away, as the ansible config system can already handle this.

rrey committed:
Implement get_option and set_options

rrey commented:
@sivel, I've tested locally and things are working like a charm, I don't understand the CI failure.

I have another question though, when I run the a playbook I get a deprecation warning:

```
[DEPRECATION WARNING]: default callback, does not support setting 'options', it will work for now,  but this will be required in the future and should be updated,  see the 2.4
 porting guide for details.. This feature will be removed in version 2.9. Deprecation warnings can be disabled by setting deprecation_warnings=False in ansible.cfg
```
But since I'm following your guideline by using get_option and set_options I'm a bit confused about what I am doing wrong.

mattclay commented:
CI failure due to traceback:

```
2018-03-24 14:28:38 Traceback (most recent call last):
2018-03-24 14:28:38   File "/root/ansible/bin/ansible-playbook", line 118, in <module>
2018-03-24 14:28:38     exit_code = cli.run()
2018-03-24 14:28:38   File "/root/ansible/lib/ansible/cli/playbook.py", line 122, in run
2018-03-24 14:28:38     results = pbex.run()
2018-03-24 14:28:38   File "/root/ansible/lib/ansible/executor/playbook_executor.py", line 89, in run
2018-03-24 14:28:38     self._tqm.load_callbacks()
2018-03-24 14:28:38   File "/root/ansible/lib/ansible/executor/task_queue_manager.py", line 210, in load_callbacks
2018-03-24 14:28:38     callback_obj.set_options()
2018-03-24 14:28:38   File "/root/ansible/lib/ansible/plugins/callback/grafana_annotations.py", line 171, in set_options
2018-03-24 14:28:38     self.grafana_api_key = self.get_option('grafana_api_key')
2018-03-24 14:28:38   File "/root/ansible/lib/ansible/plugins/callback/__init__.py", line 89, in get_option
2018-03-24 14:28:38     return self._plugin_options[k]
2018-03-24 14:28:38 KeyError: 'grafana_api_key'
```

mattclay commented:
The same failure can be seen in the unit tests as well:

https://app.shippable.com/github/ansible/ansible/runs/59064/3/tests

rrey commented:
@mattclay, I saw the traceback but I don't understand the problem.
The parameter is defined in the DOCUMENTATION string, and I can't reproduce locally.

I did not find any clear guideline on how the mechanism works, I based my patch on the slack callback as advised by @sivel. 
I confess that I don't get how the get_option/set_options mechanism works, the code from the CallbackBase did not really help.

ryancurrah commented:
I'm trying to contribute a callback plugin and I am getting the same problem locally. With all callback plugins not just mine.

rrey commented:
I managed to reproduce the issue.
I found that the `set_options` method from `CallbackBase` was not setting anything in `self._plugin_options`.

I traced the issue up to `get_plugin_options` from `ConfigManager` class, where the call to `get_configuration_definitions` receives more than just the callback name in the `name` argument:

```
> /Users/remirey/dev/ansible/lib/ansible/config/manager.py(238)get_plugin_options()
-> defs = self.get_configuration_definitions(plugin_type, name)
(Pdb) name
u'/Users/remirey/dev/ansible-callback-grafana-annotations/callback_plugins/grafana_annotations'
```

I suspect that the issue comes from the `all` method from the `PluginLoader` class where the `_update_object` method is called as follow:

```
            # load plugin config data
            if not found_in_cache:
                self._load_config_defs(basename, path)

            self._update_object(obj, name, path)
            yield obj
```

I think that the `name` argument is wrong and `basename` should be provided.
I tested the fix locally and it fixes the issue.

@bcoca, since you are the author of the code, can you advise on my analysis? Am I wrong or is it indeed the problem ?

rrey committed:
Add author parameter

bcoca commented:
@rrey the change to plugin loader seems to be required by another change to it

cc @abadger , which introduced 'basename' there

rrey commented:
@seuf this revision should fix the last comment you had.

seuf commented on lib/ansible/plugins/callback/grafana_annotations.py:
```diff
@@ -0,0 +1,261 @@
+        if self.grafana_api_key:
+            self.headers['Authorization'] = "Bearer %s" % self.grafana_api_key
+        else:
+            self.force_basic_auth = True
```
Is it possible to add a check for `grafna_user` and `grafana_password` options here ? with a warning message "You must provide a grafana_api_key or a grafana_user + grafana_password option".
The rest Looks good to me

rrey replied:
There are defaults to grafana_user and grafana_password

seuf replied:
My bad, I've missed it !

seuf reviewed (commented):
Just a minor change and it's okay

seuf reviewed (approved):
Lgtm

jhawkesworth commented:
bot_status

rrey committed:
Add the required entries in BOTMETA.yml to label the plugin community

rrey committed:
Change the 'version_added' to 2.6 now that 2.5 was released

mattclay commented on lib/ansible/plugins/callback/grafana_annotations.py:
```diff
@@ -0,0 +1,261 @@
+
+        if self.grafana_url is None:
+            self.disabled = True
+            self._display.warning('Grafana URL was not provided. The '
```
@rrey This warning shows up every time `ansible` or `ansible-playbook` is run when the plugin is not being used.

rrey replied:
ok, I'll submit a patch.

sm4rk0 replied:
I've created an issue (#38270) for this. Didn't see this comment...

bcoca replied:
this is not an issue with this callback, `set_options` should never be called unless it is specifically whitelisted
